File: src/2024/day16.rs
#[path = "../utils.rs"]
mod utils;
use utils::Point2d;

fn main() {
    // https://adventofcode.com/2024/day/16
    println!("Day 16, 2024: Reindeer Maze");

    debug_assert_eq!(7036, part1(get_example()));
    debug_assert_eq!(11048, part1(get_example2()));
    debug_assert_eq!(45, part2(get_example()));
    debug_assert_eq!(64, part2(get_example2()));

    let p1 = part1(get_input());
    println!("  Part 1: {}", p1);
    assert_eq!(88416, p1);

    let p2 = part2(get_input());
    println!("  Part 2: {}", p2);
    assert_eq!(442, p2);
}

#[derive(Clone, Debug, Eq)]
struct Pos {
    p: Point2d,
    dir: Point2d,
}

impl std::hash::Hash for Pos {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.p.hash(state)
    }
}

impl PartialEq for Pos {
    fn eq(&self, other: &Self) -> bool {
        self.p == other.p
    }
}

impl Pos {
    fn get_cost(&self, new_dir: Point2d) -> i32 {
        let cost: i32 = new_dir.x.abs() + new_dir.y.abs();

        if self.dir == new_dir {
            return cost;
        }

        cost + 1000
    }
}

struct Grid {
    map: std::collections::HashMap<Point2d, u8>,
    _width: u32,
    _height: u32,
    start: Point2d,
    end: Point2d,
}

impl Grid {
    fn parse(input: &'static str) -> Grid {
        let mut map = std::collections::HashMap::<Point2d, u8>::new();
        let mut w: u32 = 0;
        let mut h: u32 = 0;
        let mut start = Point2d::new(0, 0);
        let mut end = Point2d::new(0, 0);
        for (y, line) in input.lines().enumerate() {
            for (x, c) in line.bytes().enumerate() {
                w = std::cmp::max(w, (x + 1) as u32);
                let p = Point2d::new(x as i32, y as i32);

                let c = match c {
                    b'E' => {
                        end = p;
                        b'.'
                    }
                    b'S' => {
                        start = p;
                        b'.'
                    }
                    _ => c,
                };
                map.insert(p, c);
            }
            h = std::cmp::max(h, (y + 1) as u32);
        }

        Grid {
            map: map,
            _width: w,
            _height: h,
            start: start,
            end: end,
        }
    }

    fn get_neighbors(&self, p: &Pos) -> Vec<(Pos, i32)> {
        let mut result = vec![];
        let dirs = vec![Point2d::NORTH, Point2d::SOUTH, Point2d::EAST, Point2d::WEST];

        for new_dir in dirs {
            if p.dir == new_dir * -1 {
                continue;
            }
            let new_pos = p.p + new_dir;
            if let Some(b'.') = self.map.get(&new_pos) {
                result.push((
                    Pos {
                        p: new_pos,
                        dir: new_dir,
                    },
                    p.get_cost(new_dir),
                ));
            }
        }

        result
    }

    fn _print(&self, path: &Vec<Pos>) {
        for y in 0..self._height {
            for x in 0..self._width {
                let pt = Point2d::new(x as i32, y as i32);

                if path.iter().any(|p| p.p == pt) {
                    eprint!("O");
                } else {
                    eprint!(
                        "{}",
                        char::from_u32(*self.map.get(&pt).unwrap() as u32).unwrap()
                    );
                }
            }

            eprintln!();
        }
    }
}

fn part1(input: &'static str) -> i32 {
    let grid = Grid::parse(input);

    let result = pathfinding::directed::yen::yen(
        &Pos {
            p: grid.start,
            dir: Point2d::EAST,
        },
        |p| grid.get_neighbors(p),
        |p| p.p == grid.end,
        1,
    );

    result[0].1
}

fn part2(input: &'static str) -> u64 {
    let grid = Grid::parse(input);

    let result = pathfinding::directed::yen::yen(
        &Pos {
            p: grid.start,
            dir: Point2d::EAST,
        },
        |p| grid.get_neighbors(p),
        |p| p.p == grid.end,
        6,
    );

    let mut points = std::collections::HashSet::<Point2d>::new();
    let best_cost = result[0].1;

    for (path, cost) in result {
        if best_cost != cost {
            break;
        }

        for p in &path {
            points.insert(p.p);
        }
    }

    points.len() as u64
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2024/day16.txt").trim_end()
}

fn get_example() -> &'static str {
    "###############
#.......#....E#
#.#.###.#.###.#
#.....#.#...#.#
#.###.#####.#.#
#.#.#.......#.#
#.#.#####.###.#
#...........#.#
###.#.#####.#.#
#...#.....#.#.#
#.#.#.###.#.#.#
#.....#...#.#.#
#.###.#.#.#.#.#
#S..#.....#...#
###############"
}

fn get_example2() -> &'static str {
    "#################
#...#...#...#..E#
#.#.#.#.#.#.#.#.#
#.#.#.#...#...#.#
#.#.#.#.###.#.#.#
#...#.#.#.....#.#
#.#.#.#.#.#####.#
#.#...#.#.#.....#
#.#.#####.#.###.#
#.#.#.......#...#
#.#.###.#####.###
#.#.#...#.....#.#
#.#.#.#####.###.#
#.#.#.........#.#
#.#.#.#########.#
#S#.............#
#################"
}

File: src/2024/day16.zig
const std = @import("std");
const utils = @import("utils");

pub fn main() !void {
    // https://adventofcode.com/2024/day/16
    utils.printTitle(2024, 16, "Reindeer Maze");

    const m = utils.Monitor.init();
    defer m.deinit();
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const ally = gpa.allocator();
    const input = comptime utils.trimInput(@embedFile("input"));

    var grid = try Grid.init(ally, input);
    defer grid.deinit();

    const p1 = try grid.part1();
    utils.printAnswer(1, p1);
    std.debug.assert(88416 == p1);

    const p2 = try grid.part2();
    utils.printAnswer(2, p2);
    std.debug.assert(442 == p2);
}

const Int = i32;
const Point2d = utils.Point2d(Int);

const Elem = struct {
    pos: Point2d,
    dir: Point2d,
};

const Node = struct {
    e: Elem,
    cost: Int,
};

const dirs = [_]Point2d{
    Point2d.NORTH,
    Point2d.SOUTH,
    Point2d.EAST,
    Point2d.WEST,
};

const Grid = struct {
    ally: std.mem.Allocator,
    map: std.AutoHashMap(Point2d, u8),
    start: Point2d,
    end: Point2d,
    width: Int,
    height: Int,
    cacheForward: std.AutoHashMap(Elem, Int),
    cacheBackward: std.AutoHashMap(Elem, Int),
    best: ?Int = null,

    fn init(ally: std.mem.Allocator, input: []const u8) !Grid {
        var map = std.AutoHashMap(Point2d, u8).init(ally);

        var w: Int = 0;
        var lineIt = std.mem.tokenizeAny(u8, input, "\r\n");
        var start: Point2d = undefined;
        var end: Point2d = undefined;
        var y: Int = 0;
        while (lineIt.next()) |line| : (y += 1) {
            for (0.., line) |x, c| {
                const p = Point2d{ .x = @intCast(x), .y = y };

                const b = switch (c) {
                    'S' => blk: {
                        start = p;
                        break :blk '.';
                    },
                    'E' => blk: {
                        end = p;
                        break :blk '.';
                    },
                    else => c,
                };

                try map.put(p, b);
            }

            w = @intCast(line.len);
        }

        return .{
            .ally = ally,
            .map = map,
            .start = start,
            .end = end,
            .width = w,
            .height = y,
            .cacheForward = std.AutoHashMap(Elem, Int).init(ally),
            .cacheBackward = std.AutoHashMap(Elem, Int).init(ally),
        };
    }

    fn deinit(self: *@This()) void {
        self.map.deinit();
        self.cacheForward.deinit();
        self.cacheBackward.deinit();
    }

    fn shortestPathForward(self: *@This()) void {
        self.shortestPath(self.start, Point2d.EAST, self.end, false);
    }

    fn shortestPathBackward(self: *@This()) void {
        self.shortestPath(self.end, null, null, true);
    }

    fn shortestPath(
        self: *@This(),
        start: Point2d,
        startDir: ?Point2d,
        end: ?Point2d,
        comptime reverse: bool,
    ) void {
        const cmp = struct {
            fn call(_: void, a: Node, b: Node) std.math.Order {
                return std.math.order(a.cost, b.cost);
            }
        }.call;

        var q = std.PriorityQueue(Node, void, cmp).init(self.ally, {});
        defer q.deinit();

        var visited = std.AutoHashMap(Elem, void).init(self.ally);
        defer visited.deinit();

        if (startDir) |d| {
            const startNode = .{
                .e = .{ .pos = start, .dir = d },
                .cost = 0,
            };
            q.add(startNode) catch unreachable;
        } else {
            for (dirs) |d| {
                const startNode = .{
                    .e = .{ .pos = start, .dir = d },
                    .cost = 0,
                };
                q.add(startNode) catch unreachable;
            }
        }

        var cache = if (reverse) &self.cacheBackward else &self.cacheForward;

        while (q.items.len != 0) {
            const node = q.remove();

            if (!cache.contains(node.e)) {
                cache.putNoClobber(node.e, node.cost) catch unreachable;
            }

            if (end) |goal| {
                if (node.e.pos.eql(goal)) {
                    self.best = node.cost;
                    return;
                }
            }

            if (visited.fetchPut(node.e, {}) catch unreachable != null) {
                continue;
            }

            const d = node.e.dir;
            const newPos = node.e.pos.addp(d);
            if (self.map.get(newPos) == '.') {
                q.add(.{
                    .e = .{ .pos = newPos, .dir = d },
                    .cost = node.cost + 1,
                }) catch unreachable;
            }

            q.add(.{
                .e = .{ .pos = node.e.pos, .dir = d.rotate90Left() },
                .cost = node.cost + 1000,
            }) catch unreachable;

            q.add(.{
                .e = .{ .pos = node.e.pos, .dir = d.rotate90Right() },
                .cost = node.cost + 1000,
            }) catch unreachable;
        }
    }

    fn part1(self: *@This()) !Int {
        if (self.best == null) {
            self.shortestPathForward();
        }
        return self.best orelse 0;
    }

    fn getCost(self: *@This(), comptime reverse: bool, p: Point2d, d: Point2d) Int {
        if (reverse) {
            const elm = Elem{ .pos = p, .dir = d.neg() };
            return self.cacheBackward.get(elm) orelse 0;
        }

        const elm = Elem{ .pos = p, .dir = d };
        return self.cacheForward.get(elm) orelse 0;
    }

    fn part2(self: *@This()) !Int {
        if (self.best == null) {
            self.shortestPathForward();
        }

        self.shortestPathBackward();
        var points = std.AutoHashMap(Point2d, void).init(self.ally);
        defer points.deinit();

        var it = self.map.keyIterator();
        while (it.next()) |p| {
            for (dirs) |d| {
                if (self.getCost(false, p.*, d) + self.getCost(true, p.*, d) == self.best) {
                    try points.put(p.*, {});
                }
            }
        }

        return @intCast(points.count());
    }
};

test "parts 1,2" {
    const ally = std.testing.allocator;
    const example1 =
        \\###############
        \\#.......#....E#
        \\#.#.###.#.###.#
        \\#.....#.#...#.#
        \\#.###.#####.#.#
        \\#.#.#.......#.#
        \\#.#.#####.###.#
        \\#...........#.#
        \\###.#.#####.#.#
        \\#...#.....#.#.#
        \\#.#.#.###.#.#.#
        \\#.....#...#.#.#
        \\#.###.#.#.#.#.#
        \\#S..#.....#...#
        \\###############
    ;
    const example2 =
        \\#################
        \\#...#...#...#..E#
        \\#.#.#.#.#.#.#.#.#
        \\#.#.#.#...#...#.#
        \\#.#.#.#.###.#.#.#
        \\#...#.#.#.....#.#
        \\#.#.#.#.#.#####.#
        \\#.#...#.#.#.....#
        \\#.#.#####.#.###.#
        \\#.#.#.......#...#
        \\#.#.###.#####.###
        \\#.#.#...#.....#.#
        \\#.#.#.#####.###.#
        \\#.#.#.........#.#
        \\#.#.#.#########.#
        \\#S#.............#
        \\#################
    ;

    var grid1 = try Grid.init(ally, example1);
    defer grid1.deinit();
    try std.testing.expectEqual(7036, try grid1.part1());
    try std.testing.expectEqual(45, try grid1.part2());

    var grid2 = try Grid.init(ally, example2);
    defer grid2.deinit();
    try std.testing.expectEqual(11048, try grid2.part1());
    try std.testing.expectEqual(64, try grid2.part2());
}
